### ISSUE TITLE
Uncaught promise 

### ISSUE THREAD

**Issue by ancmikesmullin** [elastic/apm-agent-nodejs#1518]
**Describe the bug**

Uncaught promise!

There is a rule in Node.JS that you must surround all `new Promises` with a `catch`. This is typically done using `try ... catch` inbetween `async` `await` syntax, or using the chainable `Promise` `.catch()` interface. 

The fact that a promise was created without this structure indicates that the apm node programmer may not have understood a fundamental concept, and therefore you might also want to review all other Promise structures throughout the code, to be thorough about how these are handled.

The side-effect is that while APM is loaded, some redis errors are not catchable, therefore the application cannot recover from failure.

A secondary future effect is that the process may be forced to exit.

```
{
	"ts": "2019-11-13T21:46:58.039Z",
	"src": "internal/process/warning.js:27:3",
	"error": "(node:19157) UnhandledPromiseRejectionWarning: ReplyError: WRONGTYPE Operation against a key holding the wrong kind of value\n    at parseError (/workspace/projectX/node_modules/redis-parser/lib/parser.js:179:12)\n    at parseType (/workspace/projectX/node_modules/redis-parser/lib/parser.js:302:14)",
	"data": {}
}
{
	"ts": "2019-11-13T21:46:58.040Z",
	"src": "internal/process/warning.js:27:3",
	"error": "(node:19157) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)",
	"data": {}
}
{
	"ts": "2019-11-13T21:46:58.040Z",
	"src": "internal/process/warning.js:27:3",
	"error": "(node:19157) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.",
	"data": {}
}
```

**To Reproduce**

Steps to reproduce the behavior: 
1. Create a Hash by means necessary.
    `HSET a b c`
2. Then try to read it with the wrong command, using `ioredis` library with `elastic-apm-node` module loaded.
    `GET a`
3. Observe error thrown. While it is correct that Redis error should be thrown here, it is INCORRECT that it should be made UNCATCHABLE due to interference by the APM library, resulting in process exit.

**Expected behavior**

I expect when such an error is thrown, that I can catch and handle it. Like so:

```js
(async () => {
try {
  await redis.hget('a');
}
catch(e) {
  debugger; // should break here, but doesn't (only while APM is loaded)
}
})();
```

**Environment (please complete the following information)**

- OS: Linux
- Node.js version: v12.12.0
- APM Server version: N/A
- Agent version: 3.1.0

**How are you starting the agent? (please tick one of the boxes)**

- [x] Calling `agent.start()` directly (e.g. `require('elastic-apm-node').start(...)`)
- [ ] Requiring `elastic-apm-node/start` from within the source code
- [ ] Starting node with `-r elastic-apm-node/start`

**Additional context**

APM is nice. Plz fix. Also we're a paying elastic enterprise customer (Ancestry)

package.json
```
    "elastic-apm-node": "^3.1.0",
    "ioredis": "^4.11.2",
```

**Comment by axw** [comment]
@ancmikesmullin thanks for filing the issue. I've been unable to reproduce the issue so far, with this program:

```js
const apm = require('elastic-apm-node').start({serviceName: 'foo'});
var Redis = require("ioredis");
var redis = new Redis();

(async () => {
  try {
    await redis.hget('a');
  } catch(e) {
    debugger;
  }
})()
```

I'm using the same versions as you noted: node v12.12.0, ioredis v4.11.2, and elastic-apm-node v3.1.0. Regardless of whether the first line is commented out or not, the breakpoint is reached.

Seems there's more to it. Are you able to provide a minimal reproducer?

**Comment by ancmikesmullin** [comment]
```js
(async () => {

const apm = require('elastic-apm-node').start({
    serviceName: 'test',
    secretToken: 'SECRET',
    serverUrl: 'URL'
});

try {
    const Redis = require('ioredis');
    const options = {
        host: 'localhost',
        port: 6379,
    };
    const client = new Redis(options);

    let trans = apm.startTransaction('test', 'msg');

    const key2 = 'a'; // this is the HSET from earlier
    const raw2 = await client.get(key2); // this is the wrong cmd

    trans.result = 'success';
    trans.end();
}
catch(e) {
    console.log('caught error no problems');
}
})();
```

output is:

```
/home/user/.nvm/versions/node/v12.12.0/bin/node /workspace/projectX/test.js
Debugger listening on ws://127.0.0.1:46323/72cc3401-04ff-47bf-993d-24737bca045b
For help, see: https://nodejs.org/en/docs/inspector
Debugger attached.
caught error no problems
(node:517) UnhandledPromiseRejectionWarning: ReplyError: WRONGTYPE Operation against a key holding the wrong kind of value
    at parseError (/workspace/projectX/node_modules/redis-parser/lib/parser.js:179:12)
    at parseType (/workspace/projectX/node_modules/redis-parser/lib/parser.js:302:14)
(node:517) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 3)
(node:517) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:517) UnhandledPromiseRejectionWarning: ReplyError: WRONGTYPE Operation against a key holding the wrong kind of value
    at parseError (/workspace/projectX/node_modules/redis-parser/lib/parser.js:179:12)
    at parseType (/workspace/projectX/node_modules/redis-parser/lib/parser.js:302:14)
(node:517) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 4)
```

**Correction:** I AM able to catch A COPY of the error, it seems. But its not the one generating this `UnhandledPromiseRejectionWarning`.

**Correction:** This bug is NOT preventing recovery from the error, but it is generating this error output in the log, and node is threatening:
> In the future, promise rejections that are not handled will terminate the Node.js process

To solve the problem, comment out the 3 lines referencing `trans`, and the `UnhandledPromiseRejectionWarning` is no longer emitted.

**Comment by ancmikesmullin** [comment]
Also notice your example uses `hget` where I am simply using `get`. This is intentionally the wrong command for the Redis data type (GET reads String values, not Hashes), to illustrate how errors are being handled.

see also:
https://redis.io/commands/hget
vs.
https://redis.io/commands/get

**Comment by axw** [comment]
Thanks for the additional info @ancmikesmullin, that helped me reproduce the issue. I've got a fix up in #1523.

**Comment by ancmikesmullin** [comment]
Thank you! :heart:

**Comment by axw** [comment]
The fix has been released in v3.2.0.
